File: k-centre.py
import timeit
import random
import cppyy
import math
cppyy.include("k-centre.cpp")
class node:
    def __init__(self) -> None:
        self.x = 0
        self.y = 0
class graph:
    def insertNode(self):
        currX = random.randint(0, 50000)
        currY = random.randint(0, 50000)
        currNode = node()
        currNode.x = currX
        currNode.y = currY
        self.nodes.append(currNode)

    def __init__(self, AMOUNT_OF_NODES) -> None:
        self.nodes = []
        self.centers = []
        for i in range(AMOUNT_OF_NODES):
            self.insertNode()

    def d(self, lhs: node, rhs: node) -> float:
        return math.sqrt((lhs.x-rhs.x)**2 + (lhs.y-rhs.y)**2)

    def distToCenters(self, currNode: node):
        min = 900000000000
        for i in self.centers:
            currentd = self.d(currNode, i)
            if currentd < min:
                min = currentd
        return min
    def insertfarthestnode(self):
        CURRd = 0
        MAXd = 0
        MAXnode = self.nodes[0]
        for i in self.nodes:
            d = self.distToCenters(i)
            if d > MAXd:
                MAXd = d
                MAXnode = i
        self.centers.append(MAXnode)

    def kcentre(self, k):
        self.centers.clear()
        self.centers.append(self.nodes[15]) # start at arbitrary node
        while len(self.centers) != k:
            self.insertfarthestnode()


def claculateTime(AMOUNT_OF_NODES, k):
    start = timeit.default_timer()
    g = graph(AMOUNT_OF_NODES)
    g.kcentre(k)
    end = timeit.default_timer()
    pythonTime = end-start

    start = timeit.default_timer()
    cppyy.gbl.driver(k, AMOUNT_OF_NODES)
    end = timeit.default_timer()
    cppTime = end-start
    return cppTime, pythonTime

if __name__ == "__main__":
    import random
    import matplotlib.pyplot as plt
    size = 100000
    ptimes = []
    ctimes = []
    flag = 1
    numRangeStart = 5000
    i = numRangeStart
    timeToRun = 5
    start = timeit.default_timer()
    k = 4 # k-centre
    while flag:
        i+= 500
        end = timeit.default_timer()
        if end-start > timeToRun:
            print(end-start, timeToRun)
            flag = 0
        ct, pt = claculateTime(i,k)
        ptimes.append((i, pt))
        ctimes.append((i, ct))

    plt.plot(*zip(*ptimes), color="purple", label = "python")
    plt.plot(*zip(*ctimes), color="red", label = "cpp")
    plt.title("finding 2-approx for 4 centers with n nodes")
    plt.ylabel('time')
    plt.xlabel('n')
    plt.legend()
    plt.show()

        

File: k-centre.cpp
#include <vector>
#include <iomanip> // time
#include <memory>
#include <iostream>
#include <cmath>
#include <algorithm>
#include <set>



#define MAX_VALUE 50000


class graph{

class node {
    public:
    float x, y; 

    node() {
        x=0; y=0;
    }
    node(node &other) {
        x=other.x; y=other.y;
    }
    node(std::shared_ptr<graph::node> other) {
        x=other->x; y=other->y;
    }
};

public:
std::vector<std::shared_ptr<node>> nodes;
std::set<std::shared_ptr<node>> centers;
float d(std::shared_ptr<node> lhs, std::shared_ptr<node> rhs) {
    return std::sqrt((lhs->x-rhs->x)*(lhs->x-rhs->x)+(lhs->y-rhs->y)*(lhs->y-rhs->y));
}

void insertNode() {
    float CURRy = (double)(rand()) / (double)(RAND_MAX/MAX_VALUE);
    float CURRx = (double)(rand()) / (double)(RAND_MAX/MAX_VALUE);
    std::shared_ptr<node> currptr(new node());
    currptr->y = CURRy;
    currptr->x = CURRx;

    this->nodes.push_back(currptr);
}
graph(int AMOUNT_OF_NODES) {
std::srand(std::time(nullptr));
    
    for (size_t i = 0; i < AMOUNT_OF_NODES; i++)
        insertNode();            
}

float distToCenters(std::shared_ptr<node> currnode) {
    float min = 900000000000; 
    for (auto i : centers) {
        float currd = d(currnode, i);
        if (currd < min) {
            min = currd; 
        }
    }
    return min;
}

void insertfarthestnode() {
    float CURRd = 0;
    float MAXd = 0;
    std::shared_ptr<node> MAXnode = nodes.at(0);

    // for each node calculate min distance to the center set
    // take the node with the maximum min distance
    for (auto i : nodes)
    {
        float d = distToCenters(i);
        if (d>MAXd) {
            MAXd = d;
            MAXnode = i;
        }
    }
    centers.insert(std::make_shared<node>(MAXnode));
}

void kcentre(int k) {
    centers.clear();
    centers.insert(nodes.at(15));
    while (centers.size()!=k) {
        insertfarthestnode();
    }
}
};

void driver(int k, int AMOUNT_OF_NODES) {
graph G(AMOUNT_OF_NODES); 
std::vector<double> x(AMOUNT_OF_NODES), y(AMOUNT_OF_NODES), xk(k), yk(k);
for (size_t i = 0; i < G.nodes.size(); i++)
{
    x.push_back(G.nodes.at(i)->x);
    y.push_back(G.nodes.at(i)->y);
}
G.kcentre(k);
for (auto i : G.centers)
{
    xk.push_back(i->x);
    yk.push_back(i->y);
}
// Print the graph: vector x, y are nodes, vector xk, xy are centers
}